### PR TITLE
Option to poll externally managed VMs on a recurring interval

### LINKED ISSUE #78: Tenant cluster VMs running on external infra are not tracked after creation
This feature allows running tenant cluster VMs on external infrastructure (external infra meaning any cluster that is not the CAPI/CAPK management cluster itself): https://github.com/kubernetes-sigs/cluster-api-provider-kubevirt/issues/32

One outstanding issue with tenant clusters deployed on external infra is that CAPI/CAPK management plane does not track underlying VMs for tenant clusters after creation. Meaning, if an underlying VM of a tenant cluster failed completely, or restarted with a different IP, the management plane would not be able to recover the node.

This problem is resolved for tenant clusters running on the management cluster with this PR (issue https://github.com/kubernetes-sigs/cluster-api-provider-kubevirt/issues/70): https://github.com/kubernetes-sigs/cluster-api-provider-kubevirt/pull/73

But we currently lack the resolution for tenant clusters running on external infra.

Please investigate.

### PR DESCRIPTION
**Pull request by davidvossel**
Fixes #78

Externally manged VMs, meaning VMs which are not on the same local cluster that capk is running on, are not watched by informers. Instead we need to periodically wake up the reconcile loop to detect changes to these VMs. This PR adds an option that allows setting the interval the kubevirtMachine controller wakes up at for externally manged VMs

```release-note
Adds the external-cluster-sync-period cli option for specifying how often to poll externally managed VMs.
```

### CODE REVIEW

File: controllers/kubevirtmachine_controller.go [hunk 1]
```diff
@@ -339,6 +341,17 @@ func (r *KubevirtMachineReconciler) reconcileNormal(ctx *context.MachineContext)
...
 	}
 
+	// If the kubevirtMachine's VM is not managed using the default infra client
+	// provided to this pod, then we are not watching the object using an informer.
+	// Use the VM poll interval with some jitter to ensure the reconcile for this VM
+	// wakes up periodically. This is important to detect state changes such as an
+	// IP change
+	if ctx.KubevirtMachine.Spec.InfraClusterSecretRef != nil {
+		// add 10% jitter to the requeue period
+		requeueAfter := r.ExternalClusterSyncPeriod + time.Duration(rand.Intn(int(r.ExternalClusterSyncPeriod)/10))
```
Comment (qinqon) [inline controllers/kubevirtmachine_controller.go:351]: There is a Jitter function you can use for this https://pkg.go.dev/k8s.io/apimachinery/pkg/util/wait#Jitter

**Review by qinqon — CHANGES_REQUESTED** [review]
Can't we create new client with informers for the external VMIs ?

**Review by pjaton — APPROVED** [review]
Thank you @davidvossel for this change and thank you for the explanation during the bi-weekly sync. This looks good to me.

File: controllers/kubevirtmachine_controller.go (hunk 1 above)
Comment (davidvossel) [reply controllers/kubevirtmachine_controller.go]: updated

**Comment by davidvossel** [comment]
> Can't we create new client with informers for the external VMIs ?

@qinqon here's my thoughts on this.

There's no restriction for how many external clusters might be in use. Each VM could technically be on a separate external cluster. This means in order to use the informer pattern, we'd have to begin book keeping what clients belong to what VMs. Then comes the informers. We'd have to book keep what informers are shared between what VMs, and dynamically create/destroy the informers as VMs are created/deleted.

We technically can do all that, but I'd prefer seeing how far tightening the poll interval for external VMs gets us first. It's possible we don't need near instantaneous VMI status, which would make polling "good enough" for us so we can avoid maintaining the more complex informer solution.

**Comment by BarthV** [comment]
🆙 .
We're impacted by this issue too.

How can I help any of you on this topic ?

**Comment by davidvossel** [comment]
/close

I'm not satisfied with this approach. It's possible we should take an approach that avoids watching VM/VMIs entirely and always polls more often. this would solve both #78 and #100 at the cost of VM/VMI reconcile responsiveness. I think this tradeoff might be worth it though

**Comment by davidvossel** [comment]
@BarthV I've taken an alternate approach here, https://github.com/kubernetes-sigs/cluster-api-provider-kubevirt/pull/216

**Pull request closed without merging**